M2/Macaulay2/packages/GKMManifolds.m2

```diff
@@ -0,0 +1,1462 @@
+    )
+
+--given a classical Lie type LT, outputs a function f where, given a list v representing
+--a nonzero vector, f outputs {v / (abs gcd v)}, that is, a root if v is a (positive)
+-- multiple of a root, and f(v) = {} otherwise
+toRoot := LT -> (
+    if member(LT,{"A","D"}) then (
+	f := v -> (
+	    non0s := select(v, i -> i != 0);
+	    if #non0s == 2 and abs first non0s == abs last non0s then {v // (abs gcd v)} else {}
+	    )
+	)
+    else if LT == "B" then (
+	f = v -> (
+	    non0s := select(v, i -> i != 0);
+	    if #non0s == 1 or (#non0s == 2 and abs first non0s == abs last non0s) then
+	    {v // (abs gcd v)}
+	    else {}
+	    )
+	)
+    else if LT == "C" then (
+	f = v -> (
+	    non0s := select(v, i -> i != 0);
+	    if #non0s == 1 then {2 * (v // (abs gcd v))}
+	    else if (#non0s == 2 and abs first non0s == abs last non0s) then {v // (abs gcd v)}
+	    else {}
+	    )
+	)
+    )
+
+
+--input: LT, d, L
+--a string LT which is "A", "B", "C", or "D", representing the classical Lie types,
+--(we follow the convention in Example 3.4 of https://arxiv.org/pdf/1904.11029.pdf
+--for the ordering of simple roots and fundamental weights)
+--an integer d, representing the dimension of the root system,
+--(we do *not* check the low d isogenies,
+--for example, A_3 = D_3 but the code will not recognized this and output different things)
+--a list L of integers each ranging from 1 to d.
+--output:
+--a generalized flag variety G/P as a T-variety with charts,
+--where G is the classical Lie group of type LT with an action of a d-torus T,
+--(except in the case of "A", we let T be the (d+1)-torus by *not* projectivizing the torus)
+--and P is the parabolic subgroup corresponding to W_([d]\L)
+--that is, subgroup of the Weyl group generated by elements *not* in L
+--the ample class is defined to be embedding of G/P coming from the representation of G
+--whose highest weight vector equals sum(L, i -> w_i) where w_i is the i-th fundamental weight
+--(except in the orthogonal case "B,D", the "half"-fundamental weights are doubled
+--that is, we are not considering the Spin representations but representations of SO(n).
+tGeneralizedFlagVariety = method()
+tGeneralizedFlagVariety(String,ZZ,List) := TVariety => (LT,d,L) -> (
+    if not member(LT,{"A","B","C","D"}) then (
+	error " the first entry must be one of \"A\", \"B\", \"C\", or \"D\" "
+	);
+    if d <= 0 then (
+	error " the second entry (dimension) must be a positive integer "
+	);
+    if not all(L, i -> 1 <= i and i <= d) then (
+	error " indices for the weights must be between 1 and the dimension (inclusive) "
+	);
+    toRootFunc := toRoot LT;
+    R := if LT == "A" then makeCharRing(d+1) else makeCharRing d;
+    extrWts := extremalWeights(LT,d,L);
+    GEdgesDouble := new MutableHashTable;
+    chrts := apply(extrWts, v -> (
+	    lambdas := {};
+	    apply(delete(v,extrWts), w -> (
+		    lambda := toRootFunc(w-v);
+		    if #lambda == 1 then (
+			lambdas = lambdas | lambda;
```

Comment:
repeatedly lengthening a list can make an algorithm order N^2 instead of order N.  Could you just collect the results in a list and then concatenate them all at the end?